File: leetcode/add-two-numbers.cpp
// https://leetcode.com/problems/add-two-numbers/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* head = l1;
        ListNode* prev = 0;

        if (l1 == 0) {
            return l2;
        }

        while(l1 || l2) {
            if (l1 && l2) {
                int sum = l1 -> val + l2 -> val + carry;
                if (sum > 9) {
                    carry = 1;
                    l1 -> val = sum - 10;
                } else {
                    l1 -> val = sum;
                    carry = 0;
                }
                prev = l1;
                l1 = l1 -> next;
                l2 = l2 -> next;
            } else if (l1) {
                int sum = l1 -> val + carry;
                if (sum > 9) {
                    carry = 1;
                    l1 -> val = sum - 10;
                } else {
                    l1 -> val = sum;
                    carry = 0;
                }
                prev = l1;
                l1 = l1 -> next;
            } else if (l2) {
                prev -> next = l2;
                l1 = l2;
                l2 = 0;
            } else {

            }
        }

        if (carry != 0) {
            ListNode* for_carry = new ListNode();
            for_carry -> val = carry;
            prev -> next = for_carry;
        }

        return head;
    }
};

File: leetcode/add-two-numbers.go
// https://leetcode.com/problems/add-two-numbers/
package main

import "fmt"

type Node struct {
	val  int
	next *Node
}

const Base int = 10

func reverse(n *Node) *Node {
	var prev *Node = nil
	var temp *Node
	for n != nil {
		temp = n.next
		n.next = prev
		prev = n
		n = temp
	}
	return prev
}

func add(x *Node, y *Node) *Node {
	if x == nil {
		return y
	}
	if y == nil {
		return x
	}

	carry := 0
	var head *Node
	var curr *Node

	for x != nil || y != nil {
		if x != nil {
			carry += x.val
			x = x.next
		}
		if y != nil {
			carry += y.val
			y = y.next
		}
		z := &Node{
			val:  carry % Base,
			next: nil,
		}
		if head != nil {
			curr.next = z
			curr = z
		} else {
			head = z
			curr = z
		}
		carry /= Base
	}

	if carry != 0 {
		z := &Node{
			val:  carry,
			next: nil,
		}
		curr.next = z
	}

	return head
}

func main() {
	x := &Node{
		val: 8,
		next: &Node{
			val: 5,
			next: &Node{
				val:  9,
				next: nil,
			},
		},
	}

	y := &Node{
		val: 7,
		next: &Node{
			val: 6,
			next: &Node{
				val: 6,
				next: &Node{
					val:  3,
					next: nil,
				},
			},
		},
	}

	z := add(x, y)

	for z != nil {
		fmt.Println(z.val)
		z = z.next
	}
}
